### ISSUE TITLE
Only call ActiveRecord if it's defined

### ISSUE THREAD

**Issue by huoxito** [DataDog/dd-trace-rb#40]
There are a few references to ActiveRecord in the code but not all rails apps uses or even load the active_record these days. Any chance we could add checks like `if defined?` or similar around ActiveRecord references in the code pls?

```
$ git g ActiveRecord
README.md
78:ActiveRecord 5.x.

lib/ddtrace/contrib/rails/active_record.rb
9:      module ActiveRecord
20:          adapter_name = ::ActiveRecord::Base.connection_config[:adapter]

lib/ddtrace/contrib/rails/framework.rb
47:          adapter_name = ::ActiveRecord::Base.connection_config[:adapter]
67:          Datadog::Contrib::Rails::ActiveRecord.instrument()
```

**Comment by palazzem** [comment]
Thank you for pointing that! We will address that in the next release (soon)!

**Comment by huoxito** [comment]
thanks @palazzem, in case it helps this helps me boot a rails app without loading activerecord https://github.com/huoxito/dd-trace-rb/compare/master...huoxito:no-active-record

**Comment by palazzem** [comment]
thank you for the code! we will follow something really similar to your approach. Will let you know when that problem is fixed and released so that you can use again our upstream.

**Comment by msaffitz** [comment]
With the proposed fix, if ActiveRecord is defined but the database connection isn't the gem will throw an `ActiveRecord::ConnectionNotEstablished` exception.  Unfortunately there are a ton of gems that require ActiveRecord (or Rails as a whole) that don't actually use it, resulting in a soft-dependency on having the ActiveRecord gem present but remain unused.

It'd be great if the fix for this issue allowed the gem to work in the presence of ActiveRecord but absence of a DB connection.

**Comment by galdor** [comment]
Regarding this last issue, what do you mean about a database connection not being defined ? If the database cannot be reached, Rails will display an error, but there won't be any uncaught exception. Would you have an example so that we can reproduce the problem ?

**Comment by msaffitz** [comment]
@galdor here's the full stack trace we're getting:

```
/Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activerecord-3.2.22.4/lib/active_record/connection_adapters/abstract/connection_specification.rb:167:in `connection_pool': ActiveRecord::ConnectionNotEstablished (ActiveRecord::ConnectionNotEstablished)
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activerecord-3.2.22.4/lib/active_record/connection_adapters/abstract/connection_specification.rb:163:in `connection_config'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/bundler/gems/dd-trace-rb-a0837748fec2/lib/ddtrace/contrib/rails/framework.rb:62:in `configure'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/bundler/gems/dd-trace-rb-a0837748fec2/lib/ddtrace.rb:51:in `block in <class:Railtie>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:34:in `call'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:34:in `execute_hook'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:43:in `block in run_load_hooks'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:42:in `each'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:42:in `run_load_hooks'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application/finisher.rb:59:in `block in <module:Finisher>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:30:in `instance_exec'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:30:in `run'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:55:in `block in run_initializers'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:54:in `each'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:54:in `run_initializers'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application.rb:136:in `initialize!'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/railtie/configurable.rb:30:in `method_missing'
	from /Users/msaffitz/Development/apptentive/config/environment.rb:5:in `<top (required)>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/dependencies.rb:251:in `require'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/dependencies.rb:251:in `block in require'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/dependencies.rb:236:in `load_dependency'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/dependencies.rb:251:in `require'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application.rb:103:in `require_environment!'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/commands.rb:40:in `<top (required)>'
	from script/rails:7:in `require'
	from script/rails:7:in `<main>'
```

**Comment by galdor** [comment]
This problem should be fixed with #50. Could you please try `master` ? If it works, I'll close the ticket.

**Comment by msaffitz** [comment]
With active record completely removed, we saw the following:

```
Detected Rails >= 3.x. Enabling auto-instrumentation for core components.
/Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/bundler/gems/dd-trace-rb-1b724b8af03c/lib/ddtrace/contrib/rails/framework.rb:97:in `auto_instrument': uninitialized constant Datadog::Contrib::Rails::ActiveRecord (NameError)
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/bundler/gems/dd-trace-rb-1b724b8af03c/lib/ddtrace.rb:52:in `block in <class:Railtie>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:34:in `call'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:34:in `execute_hook'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:43:in `block in run_load_hooks'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:42:in `each'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-3.2.22.4/lib/active_support/lazy_load_hooks.rb:42:in `run_load_hooks'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application/finisher.rb:59:in `block in <module:Finisher>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:30:in `instance_exec'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:30:in `run'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:55:in `block in run_initializers'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:54:in `each'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/initializable.rb:54:in `run_initializers'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application.rb:136:in `initialize!'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/railtie/configurable.rb:30:in `method_missing'
	from /Users/msaffitz/Development/apptentive/config/environment.rb:5:in `<top (required)>'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/application.rb:103:in `require_environment!'
	from /Users/msaffitz/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-3.2.22.4/lib/rails/commands.rb:40:in `<top (required)>'
	from script/rails:7:in `require'
	from script/rails:7:in `<main>'
```

With ActiveRecord present (but without a DB defined), we did not have any exceptions at launch.

**Comment by palazzem** [comment]
@msaffitz we have merged #62 that should address this problem. Basically it should handle ``ActiveRecord`` like before, but it ensures that the ``Datadog::Contrib::Rails::ActiveRecord`` is defined. If you can confirm, I'll close this issue. Thank you!

**Comment by palazzem** [comment]
The fix for this issue is available in the latest [0.4.0 release][1]. Closing for now, but feel free to re-open it if something is not working properly.

[1]: https://github.com/DataDog/dd-trace-rb/releases/tag/v0.4.0
